aten/src/ATen/native/RNN.cpp

```diff
@@ -0,0 +1,700 @@
+    auto hx = std::get<0>(hidden);
+    auto cx = std::get<1>(hidden);
+
+    if (input.is_cuda()) {
+      auto igates = at::matmul(input, params.w_ih.t());
+      auto hgates = at::matmul(hx, params.w_hh.t());
+      auto result = at::_thnn_fused_lstm_cell(igates, hgates, cx, params.b_ih, params.b_hh);
+      // Slice off the workspace argument (it's needed only for AD).
+      return std::make_tuple(std::get<0>(result), std::get<1>(result));
+    }
+
+    auto gates = linear(input, params.w_ih, params.b_ih) + linear(hx, params.w_hh, params.b_hh);
+    auto chunked_gates = gates.chunk(4, 1);
+
+    auto ingate = chunked_gates[0].sigmoid();
+    auto forgetgate = chunked_gates[1].sigmoid();
+    auto cellgate = chunked_gates[2].tanh();
+    auto outgate = chunked_gates[3].sigmoid();
+
+    auto cy = (forgetgate * cx) + (ingate * cellgate);
+    auto hy = outgate * cy.tanh();
+
+    return std::make_tuple(hy, cy);
+  }
+};
+
+struct GRUCell : Cell<Tensor> {
+  hidden_type operator()(const Tensor& input, const hidden_type& hidden, const CellParams& params) const override {
+    if (input.is_cuda()) {
+      auto igates = at::matmul(input, params.w_ih.t());
+      auto hgates = at::matmul(hidden, params.w_hh.t());
+      auto result = at::_thnn_fused_gru_cell(igates, hgates, hidden, params.b_ih, params.b_hh);
+      // Slice off the workspace argument (it's needed only for AD).
+      return std::get<0>(result);
+    }
+
+    auto igates = linear(input, params.w_ih, params.b_ih);
+    auto hgates = linear(hidden, params.w_hh, params.b_hh);
+    auto chunked_igates = igates.chunk(3, 1);
+    auto chunked_hgates = hgates.chunk(3, 1);
+
+    auto reset_gate = at::sigmoid(chunked_igates[0] + chunked_hgates[0]);
+    auto input_gate = at::sigmoid(chunked_igates[1] + chunked_hgates[1]);
+    auto new_gate = at::tanh(chunked_igates[2] + reset_gate * chunked_hgates[2]);
+
+    return new_gate + input_gate * (hidden - new_gate);
+  }
+};
+
+////////////////////////////////////////////////////////////////////////////////
+// LAYER IMPLEMENTATIONS
+//
+// Layers are scan-like higher-order functions, which take in cells, and
+// transform them to fuctions of signature
+//
+// (io_type input, hidden_type hidden, param_type params) -> (io_type, hidden_type)
+//
+// which can apply the cell over a sequence of inputs, and produce both a new set
+// of hidden states, as well as a concatenated output of each step.
+
+template<typename output_type, typename hidden_type>
+struct LayerOutput {
+  output_type outputs;
+  hidden_type final_hidden;
+};
+
+template<typename io_type, typename hidden_type, typename param_type>
+struct Layer {
+  using output_type = LayerOutput<io_type, hidden_type>;
+  virtual output_type operator()(const io_type& input, const hidden_type& input_hidden, const param_type& params) const = 0;
+};
+
+template<typename hidden_type>
+struct FullLayer : Layer<Tensor, hidden_type, CellParams> {
```

Comment:
Is "full" the standard nomenclature for this sort of layer?

Reply:
I don't think there's any standard nomenclature... I can change the name if you have better ideas. Essentially it's a non-packed layer, but that sounds awful 😄